README.md
# 📢 @sefinek/google-tts-api
Google TTS (Text-To-Speech) for [Node.js](https://nodejs.org) without any vulnerabilities.

![@sefinek/google-tts-api.png](https://nodei.co/npm/@sefinek/google-tts-api.png)

![@sefinek/google-tts-api.png](https://img.shields.io/npm/dt/@sefinek/google-tts-api?maxAge=3600)
![@sefinek/google-tts-api.png](https://img.shields.io/github/last-commit/sefinek24/google-tts-api)
![@sefinek/google-tts-api.png](https://img.shields.io/github/languages/code-size/sefinek24/google-tts-api)


## 📥 Installation
```bash
$ npm install --save google-tts-api
$ npm install -D typescript @types/node # Only for TypeScript
```


## 🤔 Usage
| Method              | Options (all optional)                          | Return Type                                         | Handle Long Text |
|---------------------|-------------------------------------------------|-----------------------------------------------------|:----------------:|
| `getAudioUrl`       | `lang`, `slow`, `host`                          | `string`                                            |                  |
| `getAudioBase64`    | `lang`, `slow`, `host`, `timeout`               | `Promise<string>`                                   |                  |
| `getAllAudioUrls`   | `lang`, `slow`, `host`, `splitPunct`            | `{ shortText: string; url: string; }[]`             |        ✅         |
| `getAllAudioBase64` | `lang`, `slow`, `host`, `timeout`, `splitPunct` | `Promise<{ shortText: string; base64: string; }[]>` |        ✅         |

### Options (all optional)
| Option       | Type      | Default                      | Description                                                                                                                 |
|--------------|-----------|------------------------------|-----------------------------------------------------------------------------------------------------------------------------|
| `lang`       | `string`  | en                           | See all available language code at https://cloud.google.com/speech/docs/languages                                           |
| `slow`       | `boolean` | false                        | Use the slow audio speed if set `slow` to `true`                                                                            |
| `host`       | `string`  | https://translate.google.com | You can change the `host` if the default host could not work in your region (e.g. https://translate.google.com.cn).         |
| `timeout`    | `number`  | 10000 (ms)                   | (Only for `getAudioBase64` and `getAllAudioBase64`) Set timeout for the HTTP request.                                       |
| `splitPunct` | `string`  |                              | (Only for `getAllAudioUrls` and `getAllAudioBase64`) Set the punctuation to split the long text to short text. (e.g. "，、。") |


## 💻 Examples
### 1. `getAudioUrl(text, [option])`
```js
import * as googleTTS from '@sefinek/google-tts-api'; // ES6 or TypeScript
const googleTTS = require('@sefinek/google-tts-api'); // CommonJS

// Get audio URL
const url = googleTTS.getAudioUrl('Hello World', {
  lang: 'en',
  slow: false,
  host: 'https://translate.google.com',
});
console.log(url); // https://translate.google.com/translate_tts?...
```

### 2. `getAudioBase64(text, [option])`
```js
import * as googleTTS from '@sefinek/google-tts-api'; // ES6 or TypeScript
const googleTTS = require('@sefinek/google-tts-api'); // CommonJS

// Get base64 text
googleTTS
  .getAudioBase64('Hello World', {
    lang: 'en',
    slow: false,
    host: 'https://translate.google.com',
    timeout: 10000,
  })
  .then(console.log) // base64 text
  .catch(console.error);
```

### 3. `getAllAudioUrls(text, [option])` (For text longer than 200 characters)
```js
import * as googleTTS from '@sefinek/google-tts-api'; // ES6 or TypeScript
const googleTTS = require('@sefinek/google-tts-api'); // CommonJS

const results = googleTTS.getAllAudioUrls('LONG_TEXT_...', {
  lang: 'en',
  slow: false,
  host: 'https://translate.google.com',
  splitPunct: ',.?',
});
console.log(results);
// [
//   { shortText: '...', url: '...' },
//   { shortText: '...', url: '...' },
//   ...
// ];
```

### 4. `getAllAudioBase64(text, [option])` (For text longer than 200 characters)
```js
import * as googleTTS from '@sefinek/google-tts-api'; // ES6 or TypeScript
const googleTTS = require('@sefinek/google-tts-api'); // CommonJS

googleTTS
  .getAllAudioBase64('LONG_TEXT_...', {
    lang: 'en',
    slow: false,
    host: 'https://translate.google.com',
    timeout: 10000,
    splitPunct: ',.?',
  })
  .then(console.log)
  // [
  //   { shortText: '...', base64: '...' },
  //   { shortText: '...', base64: '...' },
  //   ...
  // ];
  .catch(console.error);
```

#### [More examples](https://github.com/sefinek24/google-tts-api/tree/master/example)


## 📑 License
MIT

CHANGELOG.md
## 2.0.2 (Mar 20, 2021)
- Change default language to `en` ([#45](https://github.com/sefinek24/google-tts-api/issues/45))
- Typescript: Remove type `Language` since API doesn't fully support language codes listed in the document

## 2.0.1 (Jan 6, 2021)
- Fix the vulnerabilities by upgrading the dependencies (#42, #44)

## 2.0.0 (Dec 8, 2020)
- Add new APIs (Please see the **Break Change** below)

| Method              | Options (optional)                              | Return Type                                         | Handle Long Text |
|---------------------|-------------------------------------------------|-----------------------------------------------------|:----------------:|
| `getAudioUrl`       | `lang`, `slow`, `host`                          | `string`                                            |                  |
| `getAudioBase64`    | `lang`, `slow`, `host`, `timeout`               | `Promise<string>`                                   |                  |
| `getAllAudioUrls`   | `lang`, `slow`, `host`, `splitPunct`            | `{ shortText: string; url: string; }[]`             |        ✅         |
| `getAllAudioBase64` | `lang`, `slow`, `host`, `timeout`, `splitPunct` | `Promise<{ shortText: string; base64: string; }[]>` |        ✅         |

- Support new Google TTS API to get audio Base64 text ([#35](https://github.com/sefinek24/google-tts-api/issues/35))
- Support long text input: `getAllAudioUrls` and `getAllAudioBase64` ([#30](https://github.com/sefinek24/google-tts-api/issues/30))
- Support changing the `host` in option ([#16](https://github.com/sefinek24/google-tts-api/issues/16))
- Support Typescript
- Add dependency [axios](https://github.com/axios/axios)

### **Break Change from 0.x.x to 2.x.x**
`googleTTS()` is changed to `googleTTS.getAudioUrl()`.

```js
const googleTTS = require('@sefinek/google-tts-api');

// Before version 0.0.6
// Original googleTTS is a promise function
const url = await googleTTS('Hello World', 'en', 1);

// After version 2.0.0
// Now googleTTS is an object with 4 new methods (getAudioUrl, getAudioBase64, getAllAudioUrls, getAllAudioBase64)
// googleTTS.getAudioUrl is a non-promise function
const url = googleTTS.getAudioUrl('Hello World', {
  lang: 'en-US',
  slow: false, // speed (number) is changed to slow (boolean)
  host: 'https://translate.google.com', // allow to change the host
});
```

## 0.0.6 (Dec 5, 2020)
- `timeout` parameter is deprecated.
- Remove dependency `isomorphic-fetch`.
- Fix the change of Google Translate API ([@freddiefujiwara](https://github.com/freddiefujiwara) in [#37](https://github.com/sefinek24/google-tts-api/pull/37)). Read more in [#35](https://github.com/sefinek24/google-tts-api/issues/35)

## 0.0.5 (Nov 8, 2020)
- Upgrade the dependencies and fix the vulnerability. ([#32](https://github.com/sefinek24/google-tts-api/issues/32))
- Add retry mechanism to prevent fetching token key failed too frequently. ([#33](https://github.com/sefinek24/google-tts-api/issues/33))

## 0.0.4 (Nov 29, 2018)
- Fix the change of Google Translate API ([@ncpierson](https://github.com/ncpierson) in [#19](https://github.com/sefinek24/google-tts-api/pull/19))

## 0.0.3 (Sep 21, 2018)
- Add package-lock.lock file
- Fix the change of Google Translate API ([@ncpierson](https://github.com/ncpierson) in [#14](https://github.com/sefinek24/google-tts-api/pull/14))

## 0.0.2 (Aug 25, 2017)
- Add yarn.lock file
- If length of input text is over than 200 characters, throw a `RangeError` with error message. ([#5](https://github.com/sefinek24/google-tts-api/issues/5))


src/getAudioBase64.ts
import assertInputTypes from './assertInputTypes';
import axios from 'axios';
import splitLongText from './splitLongText';

interface Option {
  lang?: string;
  slow?: boolean;
  host?: string;
  timeout?: number;
}

/**
 * Get "Google TTS" audio base64 text
 *
 * @param {string}   text           length should be less than 200 characters
 * @param {object?}  option
 * @param {string?}  option.lang    default is "en"
 * @param {boolean?} option.slow    default is false
 * @param {string?}  option.host    default is "https://translate.google.com"
 * @param {number?}  option.timeout default is 10000 (ms)
 * @returns {Promise<string>} url
 */
export const getAudioBase64 = async (
    text: string,
    { lang = 'en', slow = false, host = 'https://translate.google.com', timeout = 10000 }: Option = {}
): Promise<string> => {
  assertInputTypes(text, lang, slow, host);

  if (typeof timeout !== 'number' || timeout <= 0) {
    throw new TypeError('timeout should be a positive number');
  }

  if (text.length > 200) {
    throw new RangeError(
        `text length (${text.length}) should be less than 200 characters. Try "getAllAudioBase64(text, [option])" for long text.`
    );
  }

  const res = await axios({
    method: 'post',
    baseURL: host,
    url: '/_/TranslateWebserverUi/data/batchexecute',
    timeout,
    data:
        'f.req=' +
        encodeURIComponent(
            JSON.stringify([
              [['jQ1olc', JSON.stringify([text, lang, slow ? true : null, 'null']), null, 'generic']],
            ])
        ),
  });

  // 1. parse audio base64 string
  let result;
  try {
    result = eval(res.data.slice(5))[0][2];
  } catch (e) {
    throw new Error(`parse response failed:\n${res.data}`);
  }

  // Check the result. The result will be null if given the lang doesn't exist
  if (!result) {
    throw new Error(`lang "${lang}" might not exist`);
  }

  // 2. continue to parse audio base64 string
  try {
    result = eval(result)[0];
  } catch (e) {
    throw new Error(`parse response failed:\n${res.data}`);
  }

  return result;
};

interface LongTextOption extends Option {
  splitPunct?: string;
}

/**
 * @typedef {object} Result
 * @property {string} shortText
 * @property {string} base64
 */

/**
 * Split the long text into multiple short text and generate audio base64 list
 *
 * @param {string}   text
 * @param {object?}  option
 * @param {string?}  option.lang        default is "en"
 * @param {boolean?} option.slow        default is false
 * @param {string?}  option.host        default is "https://translate.google.com"
 * @param {string?}  option.splitPunct  split punctuation
 * @param {number?}  option.timeout     default is 10000 (ms)
 * @return {Result[]} the list with short text and audio base64
 */
export const getAllAudioBase64 = async (
    text: string,
    {
      lang = 'en',
      slow = false,
      host = 'https://translate.google.com',
      splitPunct = '',
      timeout = 10000,
    }: LongTextOption = {}
): Promise<{ shortText: string; base64: string }[]> => {
  assertInputTypes(text, lang, slow, host);

  if (typeof splitPunct !== 'string') {
    throw new TypeError('splitPunct should be a string');
  }

  if (typeof timeout !== 'number' || timeout <= 0) {
    throw new TypeError('timeout should be a positive number');
  }

  const shortTextList = splitLongText(text, { splitPunct });
  const base64List = await Promise.all(
      shortTextList.map((shortText) => getAudioBase64(shortText, { lang, slow, host, timeout }))
  );

  // put short text and base64 text in a list
  const result: { shortText: string; base64: string }[] = [];
  for (let i = 0; i < shortTextList.length; i++) {
    const shortText = shortTextList[i];
    const base64 = base64List[i];
    result.push({ shortText, base64 });
  }

  return result;
};
